frontend/src/components/dashboard/hooks/useAutomaticGrouping.ts

```diff
@@ -0,0 +1,322 @@
+        )
+        .with({ expressionType: ExpressionNodeType.operator }, (node) =>
+          getName(node.nodeType, node.expressionType, node.operator),
+        )
+        .with({ expressionType: ExpressionNodeType.variable }, (node) =>
+          getName(node.nodeType, node.expressionType, node.name),
+        )
+        .exhaustive(),
+    )
+    .exhaustive();
+
+const cfn = (node: AstNode): AstNode[] =>
+  match(node)
+    .returnType<AstNode[]>()
+    .with(P.array({ nodeType: AstNodeType.actor }), (nodes) => nodes)
+    .with({ nodeType: AstNodeType.actor }, (node) => [
+      ...node.eventListeners,
+      ...node.functionDeclarations,
+    ])
+    .with({ nodeType: AstNodeType.eventListener }, (node) => [
+      node.action,
+      ...node.condition.parameters,
+    ])
+    .with({ nodeType: AstNodeType.statement }, (node) =>
+      match(node)
+        .returnType<AstNode[]>()
+        .with({ statementType: StatementNodeType.assignment }, (node) => [
+          node.variable,
+          node.value,
+        ])
+        .with({ statementType: StatementNodeType.condition }, (node) =>
+          node.condition
+            ? [node.condition, node.whenTrue, node.whenFalse]
+            : [node.whenTrue, node.whenFalse],
+        )
+        .with(
+          { statementType: StatementNodeType.functionCall },
+          (node) => node.arguments,
+        )
+        .with(
+          { statementType: StatementNodeType.functionDeclaration },
+          (node) => [node.body],
+        )
+        .with({ statementType: StatementNodeType.loop }, (node) =>
+          node.condition ? [node.condition, node.body] : [node.body],
+        )
+        .with(
+          { statementType: StatementNodeType.sequence },
+          (node) => node.statements,
+        )
+        .with(
+          { statementType: StatementNodeType.variableDeclaration },
+          (node) => [node.value],
+        )
+        .exhaustive(),
+    )
+    .with({ nodeType: AstNodeType.expression }, (node) =>
+      match(node)
+        .returnType<AstNode[]>()
+        .with(
+          { expressionType: ExpressionNodeType.functionCall },
+          (node) => node.arguments,
+        )
+        .with({ expressionType: ExpressionNodeType.literal }, () => [])
+        .with(
+          { expressionType: ExpressionNodeType.operator },
+          (node) => node.operands,
+        )
+        .with({ expressionType: ExpressionNodeType.variable }, () => [])
+        .exhaustive(),
+    )
+    .exhaustive();
+
+const getInputForNode = (
+  a: AstNode,
+): {
+  root: AstNode;
+  lfn: (node: AstNode) => string;
+  cfn: (node: AstNode) => AstNode[];
+} => ({
+  root: a,
+  lfn,
+  cfn,
+});
+
+const computePqDistance = (
+  a: AstNode,
+  b: AstNode,
+  p = 2,
+  q = 3,
+): Promise<number> =>
+  new Promise((resolve) =>
+    jqgram.distance(
+      getInputForNode(a),
+      getInputForNode(b),
+      { p, q },
+      (result) => resolve(result.distance),
+    ),
+  );
```

Comment:
should we consider moving this to its own tiny library file with associated unit tests? 
I think it might make sense, also later, when evaluating new automatic grouping algorithms.